File: TypeScript/ExampleTests/common/ShoppingCartString.ts
import { AddressString } from "./AddressString.js";

import * as tokens from "./tokens.js";

export class ShoppingCartString {
  static readonly DNC_STRING = "?DNC?";

  items: string;
  shipping: string;
  discount: string;
  totalPrice: string;
  shippingAddress: AddressString;
  billingAddress: AddressString;

  constructor(items: string = "", shipping: string = "", discount: string = "", totalPrice: string = "", shippingAddress: AddressString = new AddressString(), billingAddress: AddressString = new AddressString()) {
    this.items = items;
    this.shipping = shipping;
    this.discount = discount;
    this.totalPrice = totalPrice;
    this.shippingAddress = shippingAddress;
    this.billingAddress = billingAddress;
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text: string): ShoppingCartString {
    const parts = tokens.require_(text, 6, "ShoppingCart");
    return new ShoppingCartString(
      parts[0],
      parts[1],
      parts[2],
      parts[3],
      AddressString.fromText(parts[4]),
      AddressString.fromText(parts[5])
    );
  }

  static fromList(values: Iterable<string>): ShoppingCartString {
    const v = Array.from(values);
    const r = new ShoppingCartString();
    r.items = v[0] ?? "";
    r.shipping = v[1] ?? "";
    r.discount = v[2] ?? "";
    r.totalPrice = v[3] ?? "";
    return r;
  }

  toString(): string {
    return `Items=${this.items}, Shipping=${this.shipping}, Discount=${this.discount}, TotalPrice=${this.totalPrice}, ShippingAddress=${this.shippingAddress}, BillingAddress=${this.billingAddress}`;
  }

  equals(other: ShoppingCartString): boolean {
    return (this.items === ShoppingCartString.DNC_STRING || other.items === ShoppingCartString.DNC_STRING || this.items === other.items)
      && (this.shipping === ShoppingCartString.DNC_STRING || other.shipping === ShoppingCartString.DNC_STRING || this.shipping === other.shipping)
      && (this.discount === ShoppingCartString.DNC_STRING || other.discount === ShoppingCartString.DNC_STRING || this.discount === other.discount)
      && (this.totalPrice === ShoppingCartString.DNC_STRING || other.totalPrice === ShoppingCartString.DNC_STRING || this.totalPrice === other.totalPrice)
      && this.shippingAddress.equals(other.shippingAddress)
      && this.billingAddress.equals(other.billingAddress);
  }

  static equalLists(a: ShoppingCartString[], b: ShoppingCartString[]): boolean {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]!));
  }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/ShoppingCartString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class ShoppingCartString
    {
        public string items;
        public string shipping;
        public string discount;
        public string totalPrice;
        public AddressString shippingAddress;
        public AddressString billingAddress;

        public ShoppingCartString(string items, string shipping, string discount, string totalPrice, AddressString shippingAddress, AddressString billingAddress)
        {
            this.items = items;
            this.shipping = shipping;
            this.discount = discount;
            this.totalPrice = totalPrice;
            this.shippingAddress = shippingAddress;
            this.billingAddress = billingAddress;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static ShoppingCartString FromText(string text)
        {
            var parts = Tokens.Require(text, 6, "ShoppingCart");
            return new ShoppingCartString(parts[0], parts[1], parts[2], parts[3], AddressString.FromText(parts[4]), AddressString.FromText(parts[5]));
        }

        public ShoppingCartString(string text)
        {
            var parsed = FromText(text);
            this.items = parsed.items;
            this.shipping = parsed.shipping;
            this.discount = parsed.discount;
            this.totalPrice = parsed.totalPrice;
            this.shippingAddress = parsed.shippingAddress;
            this.billingAddress = parsed.billingAddress;
        }

        public ShoppingCartTyped ToShoppingCartTyped()
        {
            return new ShoppingCartTyped(
                new List<OrderItemTyped>(),
                new Dollar(this.shipping),
                new Dollar(this.discount),
                new Dollar(this.totalPrice),
                this.shippingAddress.ToAddressTyped(),
                this.billingAddress.ToAddressTyped()
            );
        }

        public override string ToString()
        {
            return Tokens.Token(items) + " " + Tokens.Token(shipping) + " " + Tokens.Token(discount) + " " + Tokens.Token(totalPrice) + " " + Tokens.Nested(shippingAddress == null ? "" : shippingAddress.ToString()) + " " + Tokens.Nested(billingAddress == null ? "" : billingAddress.ToString());
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not ShoppingCartString other) return false;
            return (DNCString == this.items || DNCString == other.items || object.Equals(this.items, other.items))
                && (DNCString == this.shipping || DNCString == other.shipping || object.Equals(this.shipping, other.shipping))
                && (DNCString == this.discount || DNCString == other.discount || object.Equals(this.discount, other.discount))
                && (DNCString == this.totalPrice || DNCString == other.totalPrice || object.Equals(this.totalPrice, other.totalPrice))
                && object.Equals(this.shippingAddress, other.shippingAddress)
                && object.Equals(this.billingAddress, other.billingAddress);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.items);
            h.Add(this.shipping);
            h.Add(this.discount);
            h.Add(this.totalPrice);
            h.Add(this.shippingAddress);
            h.Add(this.billingAddress);
            return h.ToHashCode();
        }
    }
}
